File: src/storage/file_builder.py
"""
FileBuilder - Base Builder for Atomic File Operations

Implements the builder pattern with:
- Fluent API (with_* methods return self)
- Atomic writes (write to temp file, then rename)
- Backup and rollback support
- Configurable encoding

Usage:
    builder = FileBuilder()
    path = (
        builder
        .with_path("/path/to/file.txt")
        .with_content("Hello, World!")
        .with_encoding("utf-8")
        .with_backup(True)
        .build()
    )

    # Rollback if needed
    builder.rollback()
"""

import logging
import os
import shutil
import tempfile
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Optional, Callable, Any

logger = logging.getLogger("FileBuilder")


class FileBuilderError(Exception):
    """Exception raised for file builder errors"""
    pass


@dataclass
class FileBuilderState:
    """State container for FileBuilder"""
    path: Optional[Path] = None
    content: Optional[str] = None
    encoding: str = "utf-8"
    create_parents: bool = True
    backup_enabled: bool = False
    backup_path: Optional[Path] = None
    last_written_path: Optional[Path] = None
    temp_suffix: str = ".tmp"
    on_success: Optional[Callable[[Path], Any]] = None
    on_error: Optional[Callable[[Exception], Any]] = None


class FileBuilder:
    """
    Base file builder with atomic write support.

    Features:
    - Fluent builder pattern
    - Atomic writes via temp file + rename
    - Optional backup before overwrite
    - Rollback support
    - Configurable callbacks
    """

    def __init__(self):
        self._state = FileBuilderState()

    def with_path(self, path: str | Path) -> "FileBuilder":
        """
        Set the target file path.

        Args:
            path: Target file path (string or Path object)

        Returns:
            Self for method chaining
        """
        self._state.path = Path(path)
        return self

    def with_content(self, content: str) -> "FileBuilder":
        """
        Set the file content.

        Args:
            content: Text content to write

        Returns:
            Self for method chaining
        """
        self._state.content = content
        return self

    def with_encoding(self, encoding: str) -> "FileBuilder":
        """
        Set the file encoding.

        Args:
            encoding: Character encoding (default: utf-8)

        Returns:
            Self for method chaining
        """
        self._state.encoding = encoding
        return self

    def with_backup(self, enabled: bool = True) -> "FileBuilder":
        """
        Enable or disable backup before overwrite.

        Args:
            enabled: Whether to create backup

        Returns:
            Self for method chaining
        """
        self._state.backup_enabled = enabled
        return self

    def with_create_parents(self, enabled: bool = True) -> "FileBuilder":
        """
        Enable or disable parent directory creation.

        Args:
            enabled: Whether to create parent directories

        Returns:
            Self for method chaining
        """
        self._state.create_parents = enabled
        return self

    def with_temp_suffix(self, suffix: str) -> "FileBuilder":
        """
        Set the temporary file suffix.

        Args:
            suffix: Suffix for temp files (default: .tmp)

        Returns:
            Self for method chaining
        """
        self._state.temp_suffix = suffix
        return self

    def with_on_success(self, callback: Callable[[Path], Any]) -> "FileBuilder":
        """
        Set success callback.

        Args:
            callback: Function called with final path on success

        Returns:
            Self for method chaining
        """
        self._state.on_success = callback
        return self

    def with_on_error(self, callback: Callable[[Exception], Any]) -> "FileBuilder":
        """
        Set error callback.

        Args:
            callback: Function called with exception on error

        Returns:
            Self for method chaining
        """
        self._state.on_error = callback
        return self

    def _validate(self) -> None:
        """Validate builder state before build"""
        if self._state.path is None:
            raise FileBuilderError("Path is required. Use with_path() to set it.")
        if self._state.content is None:
            raise FileBuilderError("Content is required. Use with_content() to set it.")

    def _create_backup(self) -> Optional[Path]:
        """Create backup of existing file if enabled"""
        if not self._state.backup_enabled:
            return None

        path = self._state.path
        if path is None or not path.exists():
            return None

        # Generate backup path
        timestamp = datetime.now().strftime("%Y%m%d%H%M%S")
        backup_name = f"{path.stem}.{timestamp}.bak{path.suffix}"
        backup_path = path.parent / backup_name

        try:
            shutil.copy2(path, backup_path)
            logger.debug(f"Created backup: {backup_path}")
            return backup_path
        except Exception as e:
            logger.warning(f"Failed to create backup: {e}")
            return None

    def _atomic_write(self, path: Path, content: str, encoding: str) -> None:
        """
        Write content atomically using temp file + rename.

        This ensures that the file is either fully written or not at all,
        preventing partial writes on system crashes or interruptions.
        """
        # Create parent directories if needed
        if self._state.create_parents:
            path.parent.mkdir(parents=True, exist_ok=True)

        # Generate temp file path in same directory (for same-filesystem rename)
        temp_name = f".{path.name}.{uuid.uuid4().hex[:8]}{self._state.temp_suffix}"
        temp_path = path.parent / temp_name

        try:
            # Write to temp file
            temp_path.write_text(content, encoding=encoding)

            # Atomic rename (works on same filesystem)
            # On Windows, we need to remove the target first if it exists
            if os.name == "nt" and path.exists():
                path.unlink()

            temp_path.rename(path)
            logger.debug(f"Atomic write completed: {path}")

        except Exception as e:
            # Clean up temp file on error
            if temp_path.exists():
                try:
                    temp_path.unlink()
                except Exception:
                    pass
            raise

    def build(self) -> Path:
        """
        Build and write the file atomically.

        Returns:
            Path to the written file

        Raises:
            FileBuilderError: If validation fails or write fails
        """
        try:
            self._validate()

            path = self._state.path
            content = self._state.content
            encoding = self._state.encoding

            # Create backup if enabled
            self._state.backup_path = self._create_backup()

            # Perform atomic write
            self._atomic_write(path, content, encoding)

            # Store last written path for rollback
            self._state.last_written_path = path

            # Call success callback
            if self._state.on_success:
                self._state.on_success(path)

            logger.info(f"File written: {path}")
            return path

        except Exception as e:
            # Call error callback
            if self._state.on_error:
                self._state.on_error(e)

            if isinstance(e, FileBuilderError):
                raise
            raise FileBuilderError(f"Failed to write file: {e}") from e

    def rollback(self) -> bool:
        """
        Rollback the last write operation.

        If a backup was created, restores from backup.
        Otherwise, deletes the written file.

        Returns:
            True if rollback succeeded, False otherwise
        """
        if self._state.last_written_path is None:
            logger.warning("No previous write to rollback")
            return False

        path = self._state.last_written_path

        try:
            if self._state.backup_path and self._state.backup_path.exists():
                # Restore from backup
                shutil.copy2(self._state.backup_path, path)
                self._state.backup_path.unlink()
                logger.info(f"Rolled back from backup: {path}")
            elif path.exists():
                # No backup, just delete
                path.unlink()
                logger.info(f"Rolled back (deleted): {path}")

            self._state.last_written_path = None
            self._state.backup_path = None
            return True

        except Exception as e:
            logger.error(f"Rollback failed: {e}")
            return False

    def reset(self) -> "FileBuilder":
        """
        Reset builder state for reuse.

        Returns:
            Self for method chaining
        """
        self._state = FileBuilderState()
        return self

    @property
    def state(self) -> FileBuilderState:
        """Get current builder state (read-only view)"""
        return self._state

File: src-ts/storage/file-builder.ts
/**
 * FileBuilder - Base Builder for Atomic File Operations
 *
 * Implements the builder pattern with:
 * - Fluent API (with* methods return this)
 * - Atomic writes (write to temp file, then rename)
 * - Backup and rollback support
 * - Configurable encoding
 *
 * Usage:
 *     const path = new FileBuilder()
 *         .withPath("/path/to/file.txt")
 *         .withContent("Hello, World!")
 *         .withEncoding("utf-8")
 *         .withBackup(true)
 *         .build();
 *
 *     // Rollback if needed
 *     builder.rollback();
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import * as os from 'node:os';
import { randomUUID } from 'node:crypto';

// ---------------------------------------------------------------------------
// Error
// ---------------------------------------------------------------------------

export class FileBuilderError extends Error {
  public readonly cause?: unknown;
  constructor(message: string, options?: { cause?: unknown }) {
    super(message);
    this.name = 'FileBuilderError';
    if (options?.cause !== undefined) {
      this.cause = options.cause;
    }
  }
}

// ---------------------------------------------------------------------------
// State interface
// ---------------------------------------------------------------------------

export interface FileBuilderState {
  path: string | null;
  content: string | null;
  encoding: string;
  createParents: boolean;
  backupEnabled: boolean;
  backupPath: string | null;
  lastWrittenPath: string | null;
  tempSuffix: string;
  onSuccess: ((filePath: string) => void) | null;
  onError: ((error: Error) => void) | null;
}

function defaultState(): FileBuilderState {
  return {
    path: null,
    content: null,
    encoding: 'utf-8',
    createParents: true,
    backupEnabled: false,
    backupPath: null,
    lastWrittenPath: null,
    tempSuffix: '.tmp',
    onSuccess: null,
    onError: null,
  };
}

// ---------------------------------------------------------------------------
// FileBuilder
// ---------------------------------------------------------------------------

export class FileBuilder {
  protected _state: FileBuilderState;

  constructor() {
    this._state = defaultState();
  }

  // -- Fluent setters -----------------------------------------------------

  withPath(filePath: string): this {
    this._state.path = filePath;
    return this;
  }

  withContent(content: string): this {
    this._state.content = content;
    return this;
  }

  withEncoding(encoding: string): this {
    this._state.encoding = encoding;
    return this;
  }

  withBackup(enabled: boolean = true): this {
    this._state.backupEnabled = enabled;
    return this;
  }

  withCreateParents(enabled: boolean = true): this {
    this._state.createParents = enabled;
    return this;
  }

  withTempSuffix(suffix: string): this {
    this._state.tempSuffix = suffix;
    return this;
  }

  withOnSuccess(callback: (filePath: string) => void): this {
    this._state.onSuccess = callback;
    return this;
  }

  withOnError(callback: (error: Error) => void): this {
    this._state.onError = callback;
    return this;
  }

  // -- Validation ---------------------------------------------------------

  protected _validate(): void {
    if (this._state.path === null) {
      throw new FileBuilderError('Path is required. Use withPath() to set it.');
    }
    if (this._state.content === null) {
      throw new FileBuilderError('Content is required. Use withContent() to set it.');
    }
  }

  // -- Backup -------------------------------------------------------------

  protected _createBackup(): string | null {
    if (!this._state.backupEnabled) {
      return null;
    }

    const filePath = this._state.path;
    if (filePath === null || !fs.existsSync(filePath)) {
      return null;
    }

    const dir = path.dirname(filePath);
    const ext = path.extname(filePath);
    const base = path.basename(filePath, ext);

    const timestamp = new Date()
      .toISOString()
      .replace(/[-:T]/g, '')
      .replace(/\..+$/, '')
      .replace(/Z$/, '');

    const backupName = `${base}.${timestamp}.bak${ext}`;
    const backupPath = path.join(dir, backupName);

    try {
      fs.copyFileSync(filePath, backupPath);
      return backupPath;
    } catch {
      return null;
    }
  }

  // -- Atomic write -------------------------------------------------------

  protected _atomicWrite(
    filePath: string,
    content: string,
    encoding: string,
  ): void {
    const dir = path.dirname(filePath);

    // Create parent directories if needed
    if (this._state.createParents) {
      fs.mkdirSync(dir, { recursive: true });
    }

    // Generate temp file path in same directory (for same-filesystem rename)
    const base = path.basename(filePath);
    const uuid8 = randomUUID().split('-')[0]; // first 8 hex chars
    const tempName = `.${base}.${uuid8}${this._state.tempSuffix}`;
    const tempPath = path.join(dir, tempName);

    try {
      // Write to temp file
      fs.writeFileSync(tempPath, content, encoding as BufferEncoding);

      // On Windows, remove the target first if it exists
      if (os.platform() === 'win32' && fs.existsSync(filePath)) {
        fs.unlinkSync(filePath);
      }

      // Atomic rename (works on same filesystem)
      fs.renameSync(tempPath, filePath);
    } catch (err) {
      // Clean up temp file on error
      try {
        if (fs.existsSync(tempPath)) {
          fs.unlinkSync(tempPath);
        }
      } catch {
        // Swallow cleanup errors -- the original error is more important
      }
      throw err;
    }
  }

  // -- Build --------------------------------------------------------------

  build(): string {
    try {
      this._validate();

      const filePath = this._state.path!;
      const content = this._state.content!;
      const encoding = this._state.encoding;

      // Create backup if enabled
      this._state.backupPath = this._createBackup();

      // Perform atomic write
      this._atomicWrite(filePath, content, encoding);

      // Store last written path for rollback
      this._state.lastWrittenPath = filePath;

      // Call success callback
      if (this._state.onSuccess) {
        this._state.onSuccess(filePath);
      }

      return filePath;
    } catch (err) {
      // Call error callback
      if (this._state.onError) {
        this._state.onError(
          err instanceof Error ? err : new Error(String(err)),
        );
      }

      if (err instanceof FileBuilderError) {
        throw err;
      }
      throw new FileBuilderError(
        `Failed to write file: ${err instanceof Error ? err.message : String(err)}`,
        { cause: err },
      );
    }
  }

  // -- Rollback -----------------------------------------------------------

  rollback(): boolean {
    if (this._state.lastWrittenPath === null) {
      return false;
    }

    const filePath = this._state.lastWrittenPath;

    try {
      if (
        this._state.backupPath !== null &&
        fs.existsSync(this._state.backupPath)
      ) {
        // Restore from backup
        fs.copyFileSync(this._state.backupPath, filePath);
        fs.unlinkSync(this._state.backupPath);
      } else if (fs.existsSync(filePath)) {
        // No backup, just delete
        fs.unlinkSync(filePath);
      }

      this._state.lastWrittenPath = null;
      this._state.backupPath = null;
      return true;
    } catch {
      return false;
    }
  }

  // -- Reset --------------------------------------------------------------

  reset(): this {
    this._state = defaultState();
    return this;
  }

  // -- State getter -------------------------------------------------------

  get state(): Readonly<FileBuilderState> {
    return this._state;
  }
}
